File: Swift/AoC-Swift/AoC-Swift/Common/Day.swift
// Created on 11/23/22.

import Foundation

protocol Day: AnyObject {

  var dayNumber: Int { get }
  var year: Int { get }

  init()

  func solvePart1()

  func solvePart2()

  func part1(_ lines: [String]) -> String

  func part2(_ lines: [String]) -> String

  func part1Filename() -> String

  func part2Filename() -> String

  func readData(from filename: String) -> [String]

}

extension Day {

  func solvePart1() {
    let filename = part1Filename()
    let lines = readData(from: filename)
    let result = part1(lines)
    print("Part 1: \(result)")
  }

  func solvePart2() {
    let lines = readData(from: part2Filename())
    let result = part2(lines)
    print("Part 2: \(result)")
  }

  func readData(from filename: String) -> [String] {
    do {
      let content = try FileUtils.readLines(fromFileNamed: filename)
      return content
    } catch {
      return []
    }
  }
}

File: Java/AoC-Java/src/com/umair/aoc/common/Day.java
package com.umair.aoc.common;

import com.umair.aoc.util.FileUtils;

import java.util.List;

public abstract class Day {

  private final int dayNumber;
  private final int year;

  public Day(int day, int year) {
    this.dayNumber = day;
    this.year = year;
  }

  /**
   * Reads the file for part 1, passes it to `part1()` to solve,
   * and then returns the result.
   */
  public String solvePart1() {
    List<String> data = readData(part1Filename());
    String result = part1(data);
    return result;
  }

  /**
   * Reads the file for part 2, passes it to `part2()` to solve,
   * and then returns the result.
   */
  public String solvePart2() {
    List<String> data = readData(part2Filename());
    String result = part2(data);
    return result;
  }

  public int getDayNumber() {
    return dayNumber;
  }

  public int getYear() {
    return year;
  }

  protected abstract String part1(List<String> lines);

  protected abstract String part2(List<String> lines);

  protected abstract String part1Filename();

  protected abstract String part2Filename();

  public List<String> readData(String filename) {
    return FileUtils.readAllLinesFromFile(filename);
  }

  public String fileNameFromFileNumber(int number) {
    return FileUtils.filenameForDayAndFileNumber(this, number);
  }
}
